cabby/data/metagraph/utils.py

```diff
@@ -103,11 +108,88 @@ def update_osm_map(osm_map: Map,
   print("In update_osm_map: %d nodes in graph after adding." % (
     osm_map.nx_graph.number_of_nodes()))
 
+
+def add_conceptual_nodes_and_edges(graph: nx.Graph,
+                                   poi_map: Dict[Text, int],
+                                   wd_relations: pd.DataFrame):
+  for _, row in wd_relations.iterrows():
+    place_node_id = poi_map[row["place"]]
+    graph.add_edge(row["instanceLabel"], place_node_id)
+    graph.add_edge(place_node_id, row["instanceLabel"])
+
+def construct_human_readable_name(poi_row: pd.Series) -> Text:
+    name = ""
+    if isinstance(poi_row["name"], str):
+        name = poi_row["name"]
+    elif isinstance(poi_row["wikipedia"], str):
+        name = poi_row["wikipedia"][3:]
+    elif isinstance(poi_row["brand"], str):
+        name = poi_row["brand"]
+    elif isinstance(poi_row["shop"], str):
+        name = poi_row["shop"]
+    elif isinstance(poi_row["tourism"], str):
+        name = poi_row["tourism"]
+    elif isinstance(poi_row["amenity"], str):
+        name = poi_row["amenity"]
+    return "%s_%s" % (name, poi_row["osmid"])
+
+def convert_multidi_to_weighted_undir_graph(
+  in_graph: nx.MultiDiGraph) -> nx.Graph:
+  out_graph = nx.Graph()
+  for u, v in in_graph.edges():
+      if out_graph.has_edge(u,v):
+          out_graph[u][v]['weight'] += 1.0
+      else:
+          out_graph.add_edge(u, v, weight=1.0)
+  return out_graph
+
 def construct_metagraph(region: Region,
                         s2_level: int,
-                        base_osm_map_filepath: Text):
+                        s2_node_levels: Sequence[int],
+                        base_osm_map_filepath: Text) -> nx.Graph:
+  # Get relation data and add to existing graph.
   wd_relations = query.get_geofenced_wikidata_relations(region,
                                                         extract_qids=True)
   osm_map = Map(region, s2_level, base_osm_map_filepath)
   update_osm_map(osm_map, wd_relations)
-  return osm_map, wd_relations
\ No newline at end of file
+
+  # Construct initial metagraph and human-readable node names.
+  metagraph = convert_multidi_to_weighted_undir_graph(osm_map.nx_graph)
+  osmid_to_name = {}
+  name_to_point = {}
+  name_to_wikidata = {}
+  for _, row in osm_map.poi.iterrows():
+    name = construct_human_readable_name(row)
+    assert isinstance(name, str)
+    osmid_to_name = {row["osmid"]: name}
+    name_to_point = {name: row["geometry"]}
+    if isinstance(row["wikidata"], str):
+      name_to_wikidata[name] = row["wikidata"]
+  nx.relabel.relabel_nodes(metagraph, osmid_to_name,
+                           copy=False)
```

Comment:
Seems like this could all be on one line?

Reply:
Fixed by 32fbff65bfb0fb88b677bbec884b3583ba534b56